dll/ntdll/ldr/ldrinit.c

```diff
@@ -1455,6 +1456,80 @@ LdrpValidateImageForMp(IN PLDR_DATA_TABLE_ENTRY LdrDataTableEntry)
     UNIMPLEMENTED;
 }
 
+VOID
+NTAPI
+LdrpInitializeProcessCompat(PVOID* pOldShimData)
+{
+    static const GUID* GuidOrder[] = { &COMPAT_GUID_WIN10, &COMPAT_GUID_WIN81, &COMPAT_GUID_WIN8,
+                                       &COMPAT_GUID_WIN7, &COMPAT_GUID_VISTA };
+    static const DWORD GuidVersions[] = { WINVER_WIN10, WINVER_WIN81, WINVER_WIN8, WINVER_WIN7, WINVER_VISTA };
+
+    BYTE Buffer[sizeof(COMPATIBILITY_CONTEXT_ELEMENT) * 10 + sizeof(ACTIVATION_CONTEXT_COMPATIBILITY_INFORMATION)];
+    ACTIVATION_CONTEXT_COMPATIBILITY_INFORMATION* ContextCompatInfo;
+    SIZE_T SizeRequired;
+    NTSTATUS Status;
+    DWORD n, cur;
+
+    C_ASSERT(RTL_NUMBER_OF(GuidOrder) == RTL_NUMBER_OF(GuidVersions));
+
+    SizeRequired = sizeof(Buffer);
+    Status = RtlQueryInformationActivationContext(RTL_QUERY_ACTIVATION_CONTEXT_FLAG_NO_ADDREF,
+                                                  NULL,
+                                                  NULL,
+                                                  CompatibilityInformationInActivationContext,
+                                                  Buffer,
+                                                  sizeof(Buffer),
+                                                  &SizeRequired);
+
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("LdrpInitializeProcessCompat: Unable to query process actctx with status %x\n", Status);
+        return;
+    }
+
+    ContextCompatInfo = (ACTIVATION_CONTEXT_COMPATIBILITY_INFORMATION*)Buffer;
+    if (ContextCompatInfo->ElementCount == 0)
+        return;
+
+    for (cur = 0; cur < RTL_NUMBER_OF(GuidOrder); ++cur)
```

Comment:
This would be more efficient if this loop happened inside the ELEMENT_TYPE_OS check, then you'd completely avoid the nesting.
Probably not really relevant performance-wise though, so whatever you find easier to follow...

Reply:
This will change the logic.
I return from this function on the first one found, and we want to return the highest version present.
So first I have to check all items if I can find Win10, then check all for Win81, etc....
The case where `Type` is not `ACTCX_COMPATIBILITY_ELEMENT_TYPE_OS` is a theoretical case at the moment, but I like to write it somewhat future proof.

Reply:
Added a comment to clarify this.

Reply:
Good point.